webapp/src/js/services/db-sync.js

```diff
@@ -52,122 +46,110 @@ angular
+              // when we detect it's a 401
+              $log.error(`Denied replicating ${direction.name} remote server`, err);
+            })
+            .on('error', function(err) {
+              $log.error(`Error replicating ${direction.name} remote server`, err);
+            })
+            .on('complete', function(info) {
+              const authError = !info.ok &&
+                info.errors &&
+                info.errors.some(error => error.status === 401);
+              if (authError) {
+                Session.navigateToLogin();
+              }
+            })
+            .then(() => {
+              return; // success
+            })
+            .catch(err => {
+              $log.error(`Error replicating ${direction.name} remote server`, err);
+              return direction.name;
+            });
         });
     };
 
-    const sendUpdateForDirectedReplication = (func, direction) => {
-      return func
-        .then(() => {
-          sendUpdate({
-            direction,
-            directedReplicationStatus: 'success',
-          });
-        })
-        .catch(error => {
-          sendUpdate({
-            direction,
-            directedReplicationStatus: 'failure',
-            error,
-          });
-          return error;
-        });
-    };
-
-    var syncMeta = function() {
-      var remote = DB({ meta: true, remote: true });
-      var local = DB({ meta: true });
-      local.sync(remote);
-    };
+    const getCurrentSeq = () => DB().info().then(info => info.update_seq + '');
+    const getLastSyncSeq = () => $window.localStorage.getItem(LAST_REPLICATED_SEQ_KEY);
 
     // inProgressSync prevents multiple concurrent replications
     let inProgressSync;
+
     const sync = force => {
       if (!knownOnlineState && !force) {
         return $q.resolve();
       }
 
-      /*
-      Controllers need the status of each directed replication (directedReplicationStatus) and the
-      status of the replication as a whole (aggregateReplicationStatus).
-      */
       if (!inProgressSync) {
         inProgressSync = $q
-          .all([
-            sendUpdateForDirectedReplication(replicate('from'), 'from'),
-            sendUpdateForDirectedReplication(replicateTo(), 'to'),
-          ])
-          .then(results => {
-            const errors = _.filter(results, result => result);
-            if (errors.length > 0) {
-              $log.error('Error replicating remote server', errors);
-              sendUpdate({
-                aggregateReplicationStatus: 'required',
-                error: errors,
-              });
-              return;
-            }
-
-            syncIsRecent = true;
-            sendUpdate({ aggregateReplicationStatus: 'not_required' });
-
-            return DB().info()
-            .then(dbInfo => {
-              $window.localStorage.setItem(LAST_REPLICATED_SEQ_KEY, dbInfo.update_seq);
+          .all(DIRECTIONS.map(direction => replicate(direction)))
```

Comment:
This isn't specifically related to the changes here - but is actually a question from Enock which I couldn't answer confidently. What happens if the user fails to replicate up, but successfully replicates down. While replicating down, they get a code update. Is there any risk to the unreplicated docs? Is there any reason we'd want to confirm a successful replication TO before starting the replication FROM?

Reply:
Thanks.